gateway/conf.d/apicast.conf

```diff
@@ -95,6 +95,8 @@ location @upstream {
   proxy_pass $proxy_pass;
 
   proxy_http_version 1.1;
+  proxy_request_buffering off;
```

Comment:
I'm not sure if turning this off globally is a good idea (Slowloris DDoS attack)  May be we can have another location block  @upstream-unbufferd instead?

There is an interesting discussion on GitLab https://gitlab.com/gitlab-org/gitlab/-/issues/325095

Reply:
if we have another `location @upstream-unbufferd`, what would be the routing logic to forward to `@upstream` or `@upstream-unbuffered` ?  For requests with "small" body, no matter chunked or not, nginx will read and know the length of the body. Then send the request to upstream with "content-lenght". Thus, it does not matter `@upstream` or `@upstream-unbuffered`

Reply:
Sorry if I confused you, but what I suggest is to allow disable proxy_request_buffering per location instead of globally.
The reason are:
1. DDoS attack
2. This is a breaking change, so I think it's better to be able to configure this change per location than to go and upgrade and everything suddenly stops working.

Also, I'm just throwing ideas around, so please ignore me if I'm talking non-sense here :smile:

Reply:
So from the last meeting, we concluded that `Transfer-Encoding` header is hop-by-hop header, so the gateway is not enforced to propagate that header or transfer encoding itself. What it can be configured (exactly the same as nginx does), is the request buffering. It can be on/off.

@eguzki thinks the issue with the policy can be the order where the "request buffering" policy exists. If it is in an incorrect order, it might not work as expected. And he also thinks that we should have a single `location @upstream` and control the `proxy_request_buffering` via the Lua code.  However, something like `proxy_request_buffering $proxy_request_buffering` won't work as nginx complained on boot saying that `proxy_request_buffering` should be `on` or `off`.

My proposal is to have a  second `@upstream-unbuffered` block inside https://github.com/3scale/APIcast/blob/master/gateway/conf.d/apicast.conf
and inside the policy we set the request context, something like `request_buffered = false`
and here https://github.com/3scale/APIcast/blob/master/gateway/src/apicast/upstream.lua#L250 we can check the context and rewrite the `location_name` to `@upstream-unbuffered`